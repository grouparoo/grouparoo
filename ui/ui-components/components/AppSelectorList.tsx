import Selector, { BadgeProp } from "./Selector";
import { CardDeck } from "react-bootstrap";
import { useRouter } from "next/router";
import { Actions } from "../utils/apiData";

type SelectablePlugin = Actions.PluginsList["plugins"][number];
type SelectableApp = Actions.AppsList["apps"][number];

function isSelectableApp(item: any): item is SelectableApp {
  return item.installed === undefined;
}

export default function AppSelectorList({
  onClick,
  items,
  selectedItem,
  displayAddAppButton = false,
}: {
  onClick: Function;
  items: SelectablePlugin[] | SelectableApp[];
  selectedItem: any;
  displayAddAppButton?: boolean;
}) {
  const router = useRouter();

  return (
    <CardDeck>
      {items.map((item: SelectablePlugin | SelectableApp, idx: number) => {
        let src: string;
        let title: string;
        let subheading: string;
        let description: string;
        let className: string;
        let metaBadge: BadgeProp;
        let badges: BadgeProp[] = [];

        if (isSelectableApp(item)) {
          // these items are apps
          src = item.icon;
          title = item.name.replace("@grouparoo/", "");
          className =
            item.name === selectedItem.type
              ? "selector-list-selected"
              : "selector-list";

          if (item.provides.source) {
            badges.push({ message: "source", variant: "primary" });
          } else {
            badges.push({});
          }

          if (item.provides.destination) {
            badges.push({ message: "destination", variant: "info" });
          } else {
            badges.push({});
          }

          if (item["installed"]) {
            metaBadge = { message: "installed", variant: "warning" };
          }
        } else {
          // these items are plugins
          src = item.icon;
          title = item.name.replace("@grouparoo/", "");

          if (item.source) {
            badges.push({ message: "source", variant: "primary" });
          } else {
            badges.push({});
          }

          if (item.destination) {
            badges.push({ message: "destination", variant: "info" });
          } else {
            badges.push({});
          }

          if (item.installed) {
            metaBadge = { message: "installed", variant: "warning" };
          }

          className =
            item.name === selectedItem.name
              ? "selector-list-selected"
              : "selector-list";
        }

        return (
          <Selector
            src={src}
            title={title}
            subheading={subheading}
            description={description}
            size={120}
            className={className}
            iconClassName="card-img"
            badges={badges}
            metaBadge={metaBadge}
            onClick={() => onClick(item)}
            key={`card-${idx}`}
          />
        );
      })}

      {displayAddAppButton ? (
        <div
          style={{
            width: "200px",
            borderRadius: "4px",
            borderWidth: "1px",
            borderColor: "gray",
            borderStyle: "solid",
            margin: "15px",
            padding: "20px",
            cursor: "pointer",
          }}
          onClick={() => router.push("/app/new")}
        >
          <div className="d-flex flex-column">
            <p>Don’t see what you’re looking for?</p>
            <strong>Connect a new App</strong>
          </div>
        </div>
      ) : null}
    </CardDeck>
  );
}
