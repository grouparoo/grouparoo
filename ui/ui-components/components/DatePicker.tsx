import { Form } from "react-bootstrap";
import { useState } from "react";

export default function GrouparooDatePicker({
  selected = new Date(),
  onChange,
}: {
  selected: Date;
  onChange: Function;
}) {
  const [selectedString, setSelectedString] = useState(
    selected ? selected.toISOString().split("T")[0] : ""
  );

  const pattern = "[0-9]{4}-[0-9]{2}-[0-9]{2}";

  function formatInput(s) {
    return s.replace(/[^0-9-,]+/g, "");
  }

  return (
    <Form.Control
      type={"date"}
      value={selectedString}
      placeholder="yyyy-mm-dd"
      pattern={pattern}
      onChange={(event) => {
        const value = event.target.value;
        setSelectedString(formatInput(value));
        if (value.match(new RegExp(pattern))) {
          onChange(new Date(value));
        }
      }}
    />
  );
}
