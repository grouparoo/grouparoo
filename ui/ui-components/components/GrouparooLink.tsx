import Link from "next/link";
import { grouparooUiEdition } from "../utils/uiEdition";

const EnterpriseLink = Link;

const CommunityLink = function ({ children }) {
  return <span>{children}</span>;
};

const GrouparooLink = ["enterprise", "config"].includes(grouparooUiEdition())
  ? EnterpriseLink
  : CommunityLink;

export default GrouparooLink;
