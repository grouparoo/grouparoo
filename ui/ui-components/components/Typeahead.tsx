import { string } from "prop-types";
import { Typeahead as BootstrapTypeahead } from "react-bootstrap-typeahead";
import { Controller, Control, Path } from "react-hook-form";
import { Actions } from "../utils/apiData";

type PluginOption = (
  | Actions.SourceConnectionOptions
  | Actions.DestinationConnectionOptions
  | Actions.AppOptions
)["options"][string];

type PluginOptionValue = (
  | Actions.SourceView["source"]
  | Actions.DestinationView["destination"]
  | Actions.AppView["app"]
)["options"][string];

interface Props<T> {
  control: Control<T>;
  name: Path<T>;
  option: PluginOption;
  defaultSelected?: PluginOptionValue[];
  placeholder?: string;
  disabled?: boolean;
  required?: boolean;
  onChange?: (selected: TypeaheadOption[]) => void;
}

interface TypeaheadOption {
  key: string;
  description?: string;
}

export const FormTypeahead = <T,>({
  control,
  name,
  disabled,
  required,
  option,
  placeholder,
  defaultSelected,
  onChange,
}: Props<T>) => {
  return (
    <Controller
      control={control}
      name={name}
      render={({ field: { onChange: reactHookFormOnChange } }) => {
        return (
          <BootstrapTypeahead
            id="typeahead"
            labelKey="key"
            disabled={disabled}
            inputProps={{ required }}
            options={option?.options.map((k, idx) => {
              return {
                key: k,
                description: option?.descriptions?.[idx],
              } as TypeaheadOption;
            })}
            placeholder={placeholder}
            renderMenuItemChildren={(
              opt: TypeaheadOption,
              _props,
              idx: number
            ) => {
              return [
                <span key={`opt-${idx}-key`}>
                  {opt.key}
                  <br />
                </span>,
                <small key={`opt-${idx}-description`} className="text-small">
                  {opt.description ? (
                    <em>Description: {opt.description}</em>
                  ) : null}
                </small>,
              ];
            }}
            defaultSelected={defaultSelected}
            onChange={(selected: TypeaheadOption[]) => {
              reactHookFormOnChange(selected[0]?.key);
              onChange?.(selected);
            }}
          />
        );
      }}
    />
  );
};
