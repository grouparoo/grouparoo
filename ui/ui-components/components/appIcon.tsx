import { Image, Badge } from "react-bootstrap";

export default function AppIcon({
  src = "",
  size = 50,
  fluid = false,
  className = "",
}: {
  src?: string;
  size?: number;
  fluid?: boolean;
  className?: string;
}) {
  if (!src || src === "") {
    return (
      <div
        style={{
          padding: 10,
          textAlign: "center",
          height: fluid ? null : size,
          width: fluid ? null : size,
        }}
      >
        <br />
        <Badge>
          <em>no icon</em>
        </Badge>
      </div>
    );
  }

  if (fluid) {
    return <Image src={src} className={className} rounded fluid />;
  } else {
    return (
      <div style={{ paddingTop: fluid ? "100%" : undefined }}>
        <Image
          src={src}
          rounded
          className={className}
          style={{ height: size, width: size }}
        />
      </div>
    );
  }
}
