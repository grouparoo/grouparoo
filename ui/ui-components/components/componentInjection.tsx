import { Fragment } from "react";

// This component will wrap the entire application.
// It will be highest in the hierarchy of the page, before the <Layout> or <Page>
// You can modify the <head> of all pages by including next/head and adding script or css tags within a <head /> declaration
// See _app.tsx for the details of what is available to you in props - query, etc.

export default function Injection({ children }) {
  return <Fragment>{children}</Fragment>;
}
