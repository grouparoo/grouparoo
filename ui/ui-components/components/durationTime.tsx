import Moment from "react-moment";

export function DurationTime({ start, end }: { start: number; end: number }) {
  if (!start || !end) return <>pending</>;
  if (start === end) end++; // moment needs a non-0 delta to calculate a range
  return <Moment duration={start} date={end} />;
}
