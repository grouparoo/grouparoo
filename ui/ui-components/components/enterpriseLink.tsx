import Link from "next/link";

const EnterpriseLink = Link;

const CommunityLink = function ({ children }) {
  return <span>{children}</span>;
};

export default process.env.GROUPAROO_UI_EDITION === "enterprise" ||
process.env.GROUPAROO_UI_EDITION === "config"
  ? EnterpriseLink
  : CommunityLink;
