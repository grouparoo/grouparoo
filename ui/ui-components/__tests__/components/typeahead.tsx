import { fireEvent, render } from "@testing-library/react";
import { useForm } from "react-hook-form";
import { FormTypeahead } from "../../components/Typeahead";

describe("FormTypeahead", () => {
  it("should render", async () => {
    const Component = () => {
      const { control } = useForm();
      return (
        <FormTypeahead
          control={control}
          name="city"
          option={{
            type: "typeahead",
            options: ["Austin", "London", "Santo Domingo"],
          }}
        />
      );
    };

    const { container } = render(<Component />);
    expect(container).toMatchSnapshot("basic");
  });

  it("should render with options", async () => {
    const Component = () => {
      const { control } = useForm();
      return (
        <FormTypeahead
          control={control}
          name="city"
          option={{
            type: "typeahead",
            options: ["Austin", "London", "Santo Domingo"],
          }}
        />
      );
    };

    const { container, getByRole } = render(<Component />);

    const input = getByRole("combobox");
    fireEvent.click(input);

    expect(container).toMatchSnapshot();
  });

  it("should render with options and descriptions", async () => {
    const Component = () => {
      const { control } = useForm();
      return (
        <FormTypeahead
          control={control}
          name="city"
          option={{
            type: "typeahead",
            options: ["Austin", "London", "Santo Domingo"],
            descriptions: ["One", "Two", "Three"],
          }}
        />
      );
    };

    const { container, getByRole } = render(<Component />);

    const input = getByRole("combobox");
    fireEvent.click(input);

    expect(container).toMatchSnapshot();
  });

  it("can have a default value", async () => {
    const Component = () => {
      const { control } = useForm();
      return (
        <FormTypeahead
          control={control}
          name="city"
          defaultSelected={["Santo Domingo"]}
          option={{
            type: "typeahead",
            options: ["Austin", "Arlington", "London", "Santo Domingo"],
          }}
        />
      );
    };

    const { getByRole } = render(<Component />);
    const input = getByRole("combobox");
    expect(input).toHaveValue("Santo Domingo");
  });

  it("can be marked as required", async () => {
    const Component = () => {
      const { control } = useForm();
      return (
        <FormTypeahead
          control={control}
          name="city"
          required
          option={{
            type: "typeahead",
            options: ["Austin", "Arlington", "London", "Santo Domingo"],
          }}
        />
      );
    };

    const { getByRole } = render(<Component />);
    const input = getByRole("combobox");
    expect(input).toBeRequired();
  });

  it("should trigger custom onChange", async () => {
    const handleChange = jest.fn();
    const Component = () => {
      const { control } = useForm();
      return (
        <FormTypeahead
          control={control}
          name="city"
          onChange={handleChange}
          option={{
            type: "typeahead",
            options: ["Austin", "Arlington", "London", "Santo Domingo"],
          }}
        />
      );
    };

    const { getByRole, getByLabelText } = render(<Component />);

    const input = getByRole("combobox");
    fireEvent.change(input, { target: { value: "A" } });

    const option = getByLabelText("Arlington");
    fireEvent.click(option);

    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange).toHaveBeenCalledWith([{ key: "Arlington" }]);
  });
});
