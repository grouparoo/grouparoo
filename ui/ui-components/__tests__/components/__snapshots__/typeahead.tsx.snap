// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTypeahead should render with options 1`] = `
<div>
  <div
    class="rbt"
    style="outline: none; position: relative;"
    tabindex="-1"
  >
    <div
      style="display: flex; flex: 1; height: 100%; position: relative;"
    >
      <input
        aria-autocomplete="both"
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-owns="typeahead"
        autocomplete="off"
        class="rbt-input-main form-control rbt-input  focus"
        role="combobox"
        type="text"
        value=""
      />
      <input
        aria-hidden="true"
        class="rbt-input-hint"
        readonly=""
        style="background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.54); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%; border-style: inset inset inset inset; border-width: 2px 2px 2px 2px; line-height: normal; padding: 1px 1px 1px 1px;"
        tabindex="-1"
        value=""
      />
    </div>
    <div
      aria-label="menu-options"
      class="rbt-menu dropdown-menu show"
      id="typeahead"
      role="listbox"
      style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none; display: block; max-height: 300px; overflow: auto;"
    >
      <a
        aria-label="Austin"
        aria-selected="false"
        class="dropdown-item"
        href="#"
        id="typeahead-item-0"
        role="option"
      >
        <span>
          Austin
          <br />
        </span>
        <small
          class="text-small"
        />
      </a>
      <a
        aria-label="London"
        aria-selected="false"
        class="dropdown-item"
        href="#"
        id="typeahead-item-1"
        role="option"
      >
        <span>
          London
          <br />
        </span>
        <small
          class="text-small"
        />
      </a>
      <a
        aria-label="Santo Domingo"
        aria-selected="false"
        class="dropdown-item"
        href="#"
        id="typeahead-item-2"
        role="option"
      >
        <span>
          Santo Domingo
          <br />
        </span>
        <small
          class="text-small"
        />
      </a>
    </div>
  </div>
</div>
`;

exports[`FormTypeahead should render with options and descriptions 1`] = `
<div>
  <div
    class="rbt"
    style="outline: none; position: relative;"
    tabindex="-1"
  >
    <div
      style="display: flex; flex: 1; height: 100%; position: relative;"
    >
      <input
        aria-autocomplete="both"
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-owns="typeahead"
        autocomplete="off"
        class="rbt-input-main form-control rbt-input  focus"
        role="combobox"
        type="text"
        value=""
      />
      <input
        aria-hidden="true"
        class="rbt-input-hint"
        readonly=""
        style="background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.54); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%; border-style: inset inset inset inset; border-width: 2px 2px 2px 2px; line-height: normal; padding: 1px 1px 1px 1px;"
        tabindex="-1"
        value=""
      />
    </div>
    <div
      aria-label="menu-options"
      class="rbt-menu dropdown-menu show"
      id="typeahead"
      role="listbox"
      style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none; display: block; max-height: 300px; overflow: auto;"
    >
      <a
        aria-label="Austin"
        aria-selected="false"
        class="dropdown-item"
        href="#"
        id="typeahead-item-0"
        role="option"
      >
        <span>
          Austin
          <br />
        </span>
        <small
          class="text-small"
        >
          <em>
            Description: 
            One
          </em>
        </small>
      </a>
      <a
        aria-label="London"
        aria-selected="false"
        class="dropdown-item"
        href="#"
        id="typeahead-item-1"
        role="option"
      >
        <span>
          London
          <br />
        </span>
        <small
          class="text-small"
        >
          <em>
            Description: 
            Two
          </em>
        </small>
      </a>
      <a
        aria-label="Santo Domingo"
        aria-selected="false"
        class="dropdown-item"
        href="#"
        id="typeahead-item-2"
        role="option"
      >
        <span>
          Santo Domingo
          <br />
        </span>
        <small
          class="text-small"
        >
          <em>
            Description: 
            Three
          </em>
        </small>
      </a>
    </div>
  </div>
</div>
`;

exports[`FormTypeahead should render: basic 1`] = `
<div>
  <div
    class="rbt"
    style="outline: none; position: relative;"
    tabindex="-1"
  >
    <div
      style="display: flex; flex: 1; height: 100%; position: relative;"
    >
      <input
        aria-autocomplete="both"
        aria-expanded="false"
        aria-haspopup="listbox"
        autocomplete="off"
        class="rbt-input-main form-control rbt-input"
        role="combobox"
        type="text"
        value=""
      />
      <input
        aria-hidden="true"
        class="rbt-input-hint"
        readonly=""
        style="background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.54); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%;"
        tabindex="-1"
        value=""
      />
    </div>
  </div>
</div>
`;
